File: demo2.m
clear; clc;
%% 数据处理
data_temp = imread('./2.png');
data_2 = data_temp(135:719,219:1059,1);
data_2 = [data_2; zeros(256,841)];
path = './Screenshot_2020-03-20-01-54-08.png';
data_temp = imread(path);
test_2 = data_temp(135:719,219:1059,1);
test_2 = [test_2; zeros(256,841)];
%% 剪切剩扇形
for i = 1:841 %行
    for j = 1:841 %列
        dis = sqrt(power(i-421,2)+power(j-421,2));
        if dis < 142.8 || dis > 420.5
            test_2(i,j) = 0;
            data_2(i,j) = 0;
        end
    end
end
% test = imrotate(test_2,16,'nearest','crop');
% imshow(test_2);
% hold on;
% x = 1:420;
% y = 420 - (420-x)*abs(tan(165/180*pi));
% plot(x,y,'.');

% for i = 1:420 %行
% 	for j = 1:420 %列
%         tann = (421-j)/(421-i);
%         if tann <= tan_bigger && tann >= tan_lower
%             test(j,i) = 255;
%         end
% 	end
% end
% imshow(test);
%% 开始旋转
N = 10;
left_angle = 170;
angle_space = 160 / N;
show_range = [left_angle-angle_space+4.5, left_angle-4.5];
tan_bigger = abs(tan(show_range(1)/180*pi));
tan_lower = abs(tan(show_range(2)/180*pi));

% % 演示剪切剩最左扇形
% data_rotate = imrotate(data_2,16*0,'nearest','crop');
% data_rotate(421:end,:) = 0;
% data_rotate(:,421:end) = 0;
% for i = 1:420 %行
% 	for j = 1:420 %列
%         tann = (421-j)/(421-i);
%         if tann > tan_bigger || tann < tan_lower
%             data_rotate(j,i) = 0;
%         end
% 	end
% end
% imshow(data_rotate);

% % x = 1:421;
% % y1 = 421 - (421-x)*abs(tan(166.5/180*pi));
% % y2 = 421 - (421-x)*abs(tan(157.5/180*pi));
% % plot(x,y1,'.');
% % plot(x,y2,'.');

data_slice = zeros(420,420,3);
data_slice = uint8(data_slice);
for n = 1:N
    data_rotate = imrotate(data_2,angle_space*(n-1),'nearest','crop');
    for i = 1:420 %行
        for j = 1:420 %列
            tann = (421-j)/(421-i);
            if tann > tan_bigger || tann < tan_lower
                data_rotate(j,i) = 0;
            end
        end
    end
    data_slice(:,:,n) = data_rotate(1:420,1:420);
end
% imshow(data_slice(:,:,5));

test_slice = zeros(420,420,3);
test_slice = uint8(test_slice);
for n = 1:N
    test_rotate = imrotate(test_2,angle_space*(n-1),'nearest','crop');
    for i = 1:420 %行
        for j = 1:420 %列
            tann = (421-j)/(421-i);
            if tann > tan_bigger || tann < tan_lower
                test_rotate(j,i) = 0;
            end
        end
    end
    test_slice(:,:,n) = test_rotate(1:420,1:420);
end
% imshow(test_slice(:,:,5));

total_cha = [];
for i = 1:N
    minus = data_slice - test_slice(:,:,i);
    cha = sum(sum(minus,2),1);
    cha = reshape(cha, 1, N);
    total_cha = [total_cha;cha]
end

[c,t] = hungarian(total_cha');%每个位置放的第几块正确拼图

%% 按顺序恢复

move = zeros(1,N);
for i = 1:N
    [sorted, idx] = sort(c);
    if c(i)~=i
        move(i) = idx(i);
        t = c(i);
        c(i) = c(idx(i));
        c(idx(i)) = t;
    else
       move(i) = 0;;% 0不用移动
    end
end
disp(move);

File: HuaYuJian/XiuShanZi_python/demo2.py
# -*- coding: utf-8 -*-
from PIL import Image
import matplotlib.pyplot as plt
import numpy as np
from scipy.optimize import linear_sum_assignment
from pymouse import PyMouse
from PyQt5.QtWidgets import QApplication
from PyQt5.QtGui import *
import win32gui
import sys
import time

test_name = '4_test.png'

############ capture windows ############
hwnd = win32gui.FindWindow(None, '1')
app = QApplication(sys.argv)
screen = QApplication.primaryScreen()
im = screen.grabWindow(hwnd).toImage()

## Qimage type to numpy
size = im.size()
s = im.bits().asstring(size.width() * size.height() * im.depth() // 8)  # format 0xffRRGGBB
arr = np.fromstring(s, dtype=np.uint8).reshape((size.height(), size.width(), im.depth() // 8))
arr = arr[36:756,1:1281,0:3]

# imgs = Image.fromarray(arr)
# imgs.save("2_test.png")

data_temp = Image.open('3.png')
data_2 = np.array(data_temp)
data_2 = data_2[134:719,218:1059,1]
zeros = np.zeros((256,841))
data_2 =  np.concatenate((data_2,zeros),axis=0)

# test_temp = Image.open(test_name)
# test_2 = np.array(test_temp)
test_2 = arr
test_2 = test_2[134:719,218:1059,1]
zeros = np.zeros((256,841))
test_2 =  np.concatenate((test_2,zeros),axis=0)

## cut round area
masking_round = np.zeros((841,841))
# i->行 j->列
for i in range(841):
    for j in range(841):
        dis = np.sqrt(pow(i-419,2)+pow(j-419,2))
        if (dis>=142.8) and (dis<=420.5):
            masking_round[j,i] = 1
        else:
            masking_round[j,i] = 0
data_2 = data_2 * masking_round
test_2 = test_2 * masking_round

# plt.figure()
# show = Image.fromarray(test_2)
# plt.imshow(show)
# plt.show()
## rotate argument
N = 10
left_angle = 170
angle_space = 160 / N
show_range = [left_angle-2*angle_space+1, left_angle-angle_space-1]
tan_bigger = np.abs(np.tan(show_range[0]/180*np.pi))
tan_lower = np.abs(np.tan(show_range[1]/180*np.pi))

masking_sector = masking_round
for i in range(420):
    for j in range(420):
        tann = (420-j)/(420-i)
        if (tann>tan_bigger) or (tann<tan_lower):
            masking_sector[j,i] = 0

for i in range(420):
    masking_sector[i+420,:] = 0

for i in range(420):
    masking_sector[:,i+420] = 0

############ get sector slice ############
gray2white = 30

data_im = Image.fromarray(data_2)
data_slice = np.zeros((420,420,N))
for n in range(N):
    data_rotate = data_im.rotate(angle_space*(n-1))
    data_array = np.array(data_rotate)
    if(n%2==1):
        data_array = data_array + gray2white
    data_array = data_array * masking_sector
    data_slice[:,:,n] = data_array[0:420,0:420]

test_im = Image.fromarray(test_2)
test_slice = np.zeros((420,420,N))
for n in range(N): 
    test_rotate = test_im.rotate(angle_space*(n-1))
    test_array = np.array(test_rotate)
    test_array = test_array * masking_sector
    if(n%2==1):
        test_array = test_array + gray2white
    test_array = test_array * masking_sector
    test_slice[:,:,n] = test_array[0:420,0:420]

# plt.figure()
# show1 = np.concatenate((test_slice[:,:,3],data_slice[:,:,4]),axis=0)
# show = Image.fromarray(show1)
# plt.imshow(show)
# plt.show()

# im1 = Image.fromarray(test_slice[:,:,14])
# plt.hist(test_slice[:,:,14], 256, [0, 256])
# plt.show()


diff=np.zeros((N,N))

for n in range(N):
    for m in range(N):
        cha = data_slice[:,:,n] - test_slice[:,:,m]
        cha = np.abs(cha)
        cha = cha.sum()
        diff[m,n] = cha
    
r, c = linear_sum_assignment(diff)

# print(r,c)
# test_show = Image.fromarray(test_slice[:,:,9])
# plt.imshow(test_show)
# plt.show()

############ calculate move ############
move = [0,1,2,3,4,5,6,7,8,9]

for i in range(N):
    idx = np.argsort(c)
    if(c[i]!=i):
        t = c[i]
        c[i] = c[idx[i]]
        c[idx[i]] = t
        move[i] = idx[i]
    else:
        move[i] = -1
print(move)


############ start click ############
# m = PyMouse()

# pos_10 = [[220, 371],[264, 304],[317, 241],[382, 190],[467, 165],
# [545, 160],[629, 198],[694, 252],[750, 314],[790, 377]]

# m.click(pos_10[0][0], pos_10[0][1])
# time.sleep(1)

# def change(num1, num2):
#     m.press(pos_10[num1][0],pos_10[num1][1])
#     time.sleep(0.01)
#     m.press(pos_10[num2][0],pos_10[num2][1])
#     time.sleep(0.77)

# def change_all(move):
#     m.click(100,100)
#     time.sleep(0.009)
#     for i in range(10):
#         if(move[i]!=-1):
#             change(i,move[i])

# change_all(move)